### PR TITLE
Add paragraph motion and textobject

### LINKED ISSUE #1580: Add paragraph + indenation text object/motion
These are very simple to define (by whitespace) and often reflect logical groupings that are not part of a programming language's grammar (not Treesitter's domain).

I tend to put lines of code that are related together, and I do my best to indent things semantically. I think this applies to most programmers across all filetypes so it makes sense as language-agnostic functionality.

### PR DESCRIPTION
**Pull request by pickfire**
Also improved testing facility.

Fix #1580
Superseeds #784

Although this is not done I think it would be good to help review the new test helpers.

# TODO

- [x] prev paragraph (test)
- [x] next paragraph
- [x] inside/around paragraph
- [ ] support extend mode?

Not sure if we want to support extend mode for this even though inspiration is taken from kakoune, kakoune does not have extend mode but we have that, maybe we can make use of it to extend an additional paragraph? I didn't do it since I don't think I need it.

### CODE REVIEW

File: helix-term/src/keymap.rs [hunk 1]
```diff
@@ -607,15 +607,17 @@ impl Default for Keymaps {
                 "D" => goto_first_diag,
                 "f" => goto_prev_function,
                 "c" => goto_prev_class,
-                "p" => goto_prev_parameter,
+                "a" => goto_prev_parameter,
+                "p" => move_prev_para,
```
Comment (pickfire) [inline helix-term/src/keymap.rs:611]: cc @sudormrfbin I changed `p` to `a` for parameter because parameter are also known as argument (although the meaning is slightly different) but paragraph have no alternatives, the other reason being paragraph is a known motion in both vim and kakoune, this can keep the consistency and make it easier to learn.

**Comment by dead10ck** [comment]
I think the syntax of the test helpers warrants its own whole discussion. Should we do that in this PR, or save it for another? It's also very relevant to the integration testing branch I've been working on.

**Comment by pickfire** [comment]
> I think the syntax of the test helpers warrants its own whole discussion. Should we do that in this PR, or save it for another? It's also very relevant to the integration testing branch I've been working on.

I use that here because it's very painful to have to manually count `\n` while I was working on this, I spent a whole bunch of time counting them until I feel the pain to write it out, maybe we can change the test syntax later?

**Comment by dead10ck** [comment]
Yeah I felt that pain when writing the auto pairs tests as well, counting individual characters. I'd really like to improve the testing facilities too, but it was apparent in the chat room that there were some different ideas about how to mark selections and such, so it would be better to work out a scheme before it becomes much more difficult to change later because there are a ton of tests written with the chosen syntax. But this discussion may not be short, so I also don't want to hold up this PR.

**Comment by pickfire** [comment]
> But this discussion may not be short, so I also don't want to hold up this PR.

I rather hold up this PR and wait far that discussion, given that it is so error prone while writing index based test like I did previously, usually it's fine but now `\n` visually 2 characters became 1 character and my eyes T_T.

**Comment by dead10ck** [comment]
Okay, I'm also in favor of discussing now.

Personally I think we will need to be able to write single special ASCII characters, especially in integration tests, where, for example, there may end up being lots of tests to verify things like our tree-sitter queries in actual real source code, so I think something about the syntax will need to change from the one in this PR.

So I think in order to avoid the whole issue of escaping, it might be better to use a multi character sigil to mark selections, so that the syntax is very unlikely to conflict with any input text.

I know @archseer mentioned that kakoune uses `%()` to denote selections. I'm not sure what other syntax there is, since it doesn't seem to be documented. This might work, although we should probably extend it further so we can indicate the head, as well as primary selection, like you tried to do in this PR.

Maybe to simplify the scanner, we could do `%( )%` and `|` at the beginning or end to mark the head? So

`%(|some te)%xt` would mean `some te` is the selection, with the head at the front.

And maybe secondary selections could be `%[text]%`.

Just spit balling. Anyone have any other ideas?

Also, I'd really like to pull in [indoc](https://github.com/dtolnay/indoc) so we can use raw strings, so the input text will look much more like rendered text. I think in cases where it's more important to verify exact line endings, these can just not use indoc so you can be sure of the input.

File: helix-core/src/movement.rs [hunk 2]
```diff
@@ -149,6 +150,77 @@ fn word_move(slice: RopeSlice, range: Range, count: usize, target: WordMotionTar
     })
 }
 
+pub fn move_prev_para(slice: RopeSlice, range: Range, count: usize, behavior: Movement) -> Range {
```
Comment (anthonyjclark) [inline helix-core/src/movement.rs:153]: (Snooping around on PRs to get a feel for the code base.)

Would this not be better as `move_prev_paragraph` (emphasis on spelling out paragraph) to disambiguate from parameter? It might also be good in this PR to change all reference "parameter" to "argument," but that is more so  a matter of taste and if the character "a" ends up being used (e.g., "]a").

**Comment by archseer** [comment]
> although we should probably extend it further so we can indicate the head,

It seems to me that `%` indicates the tail: `%()` vs `()%`. For primary vs other selections we can use a different symbol, i.e. `%[]`

**Comment by dead10ck** [comment]
> > although we should probably extend it further so we can indicate the head,
> 
> It seems to me that `%` indicates the tail: `%()` vs `()%`. For primary vs other selections we can use a different symbol, i.e. `%[]`

That would make sense; however, just thinking about it from an implementation perspective, it might be tricky to implement a parser for this without the same problem with escaping. e.g. if we're testing some Rust source code

```rust
let pair = ((bar(), baz()))%;
```

How do we know where the selection starts? We'd have to do some kind of backtracking and possibly escaping or matching of other `()` pairs.

I think having both ends be two chars will make it easier to parse.

File: helix-core/src/movement.rs (hunk 2 above)
Comment (pickfire) [reply helix-core/src/movement.rs:153]: Ah, when I wrote this I made it `para` because it have the same length as `word`, but parameter was there after I rebased, I guess I will change it.

**Comment by pickfire** [comment]
Not quite the example you pointed out but I imagine this would be more common.
```python
x = (a-b)%(c-b)
```

**Comment by dead10ck** [comment]
> Not quite the example you pointed out but I imagine this would be more common. I think maybe using `#` is better since I rarely see `#()`.
> 
> ```python
> x = (a-b)%(c-b)
> ```

Yeah, this case could be easily fixed by adding whitespace, but `#()#` works just as well I think.

**Comment by archseer** [comment]
I don't have a strong opinion here so I'm okay with whatever you two settle on. I'd keep it simple and pick something "good enough" even if it has some edge cases, we can avoid those by changing the test specs.

**Comment by dead10ck** [comment]
Ok, @pickfire I'm good with `#(|)#` and `#[|]#` if you are.

**Comment by pickfire** [comment]
> Ok, @pickfire I'm good with #(|)# and #[|]# if you are.

I don't think that is a good idea since we don't know which is head and anchor. But `#[]` for primary and `#()` for others should be good enough I think.

**Comment by dead10ck** [comment]
@pickfire read [my earlier comment](https://github.com/helix-editor/helix/pull/1627#issuecomment-1047030875) for how to mark the head and anchor.

`#[|hello]# world` would indicate `hello` is the selection with the head coming before the anchor. `#[hello|]# world` would indicate `hello` with the head at the end.

**Comment by pickfire** [comment]
Wouldn't it be better to have `#[|]#` for primary given that primary selection is always needed and used more often than secondary selection, so it shouldn't have the extra shift needed?

**Comment by dead10ck** [comment]
> Wouldn't it be better to have `#[|]#` for primary given that primary selection is always needed and used more often than secondary selection, so it shouldn't have the extra shift needed on most layouts?

Sure, that sounds good to me 👍

File: helix-core/src/test.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            left.push(c);
+        }
+    }
+    let primary = match primary {
+        Some(i) => i,
+        None => panic!("missing primary `#[|]#` {s:?}"),
+    };
+    let selection = Selection::new(ranges, primary);
+    (left, selection)
+}
```
Comment (pickfire) [inline helix-core/src/test.rs:118]: This seemed quite long, not sure if changing it to work with str will improve it or not but str we need to search it multiple times compared to this way.

Comment (dead10ck) [reply helix-core/src/test.rs:118]: Just after a quick once over, this doesn't look *too* bad; it's not pretty, but it's not incomprehensible. It seems simple enough that I don't foresee a big maintenance problem, assuming we don't keep adding to the syntax, and I can't imagine how we would.

However, it is basically implementing a state machine parser by hand. I wonder if it would be overkill to pull in something like [`combine`](https://github.com/Marwes/combine) to offload the parsing. Taking a quick glance at the `Cargo.toml`, it appears that it only has one hard dependency on `memchr`, so it shouldn't be too bad as far as compile times. Just an idea though, I don't have a strong preference for this.

Comment (sudormrfbin) [reply helix-core/src/test.rs:118]: Some early returns can make it more readable:

```rust
pub fn print(s: &str) -> (String, Selection) {
    let mut primary_idx = None;
    let mut ranges = SmallVec::new();
    let mut iter = s.chars().peekable();
    let mut left = String::with_capacity(s.len());

    'outer: while let Some(c) = iter.next() {
        let start = left.len();

        if c != '#' {
            left.push(c);
            continue;
        }

        let (is_primary, close_pair) = match iter.next() {
            Some('[') => (true, ']'),
            Some('(') => (false, ')'),
            Some(ch) => {
                left.push(ch);
                continue;
            }
            None => break,
        };

        if is_primary && primary_idx.is_some() {
            panic!("primary `#[` already appeared {left:?} {s:?}");
        }

        let head_at_beg = iter.next_if_eq(&'|').is_some();

        'in_range: while let Some(c) = iter.next() {
            if !(c == close_pair && iter.peek() == Some(&'#')) {
                left.push(c);
                continue 'in_range;
            }
            if !head_at_beg {
                let prev = left.pop().unwrap();
                if prev != '|' {
                    left.push(prev);
                    left.push(c);
                    continue 'in_range;
                }
            }
            iter.next(); // skip "#"

            if is_primary {
                primary_idx = Some(ranges.len());
            }
            let (anchor, head) = match head_at_beg {
                true => (left.len(), start),
                false => (start, left.len()),
            };
            ranges.push(Range::new(anchor, head));
            continue 'outer;
        }

        if head_at_beg {
            panic!("missing end `{close_pair}#` {left:?} {s:?}");
        } else {
            panic!("missing end `|{close_pair}#` {left:?} {s:?}");
        }
    }

    let primary = match primary_idx {
        Some(i) => i,
        None => panic!("missing primary `#[|]#` {s:?}"),
    };
    let selection = Selection::new(ranges, primary);
    (left, selection)
}
```

Comment (pickfire) [reply helix-core/src/test.rs]: Thanks, applied with some minor tweaks to remove inner loop label.

File: helix-core/src/test.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        if c != '#' {
+            left.push(c);
+            continue;
+        }
+
+        let (is_primary, close_pair) = match iter.next() {
+            Some('[') => (true, ']'),
+            Some('(') => (false, ')'),
+            Some(ch) => {
+                left.push(ch);
```
[the remark this answers is not in the document]
Comment (pickfire) [reply helix-core/src/test.rs:48]: Thanks

**Comment by wurosh** [comment]
I'd just like to chime in on adding extend mode support. I personally use it with vim quite a bit, but that's beside the point. I believe it should be implemented for consistency's sake - why would motions suddenly not work when I happen to be selecting things? In addition, in obscure languages where treesitter fails/is not complete, or when editing markdown documents, there's really no good way to select an entire class/multiple paragraphs other than extend mode.

PS: Is there a reason why existing objects are not listed in the popup when one types `mi`/`ma`?

**Comment by sudormrfbin** [comment]
> PS: Is there a reason why existing objects are not listed in the popup when one types `mi`/`ma`?

Fixed in master: https://github.com/helix-editor/helix/pull/1686

**Comment by pickfire** [comment]
> I'd just like to chime in on adding extend mode support. I personally use it with vim quite a bit, but that's beside the point. I believe it should be implemented for consistency's sake - why would motions suddenly not work when I happen to be selecting things? In addition, in obscure languages where treesitter fails/is not complete, or when editing markdown documents, there's really no good way to select an entire class/multiple paragraphs other than extend mode.

Note that some parts of extend (or normal select behavior) are still inconsistent, I am looking to send fixes as I see fit.

**Comment by pickfire** [comment]
> I'd just like to chime in on adding extend mode support. I personally use it with vim quite a bit, but that's beside the point. I believe it should be implemented for consistency's sake - why would motions suddenly not work when I happen to be selecting things? In addition, in obscure languages where treesitter fails/is not complete, or when editing markdown documents, there's really no good way to select an entire class/multiple paragraphs other than extend mode.

Note that some parts of extend (or normal select behavior) are still inconsistent, I am looking to send fixes as I see fit.

**Review by archseer — APPROVED** [review]
Looks good!

**Comment by archseer** [comment]
Just needs a rebase to fix the conflict

**Review by archseer — APPROVED** [review]

**Comment by sudormrfbin** [comment]
Thanks @pickfire for working on this 🎉  Holding `x` was getting really tiring :P
